Add responsetime/uptime/connection-count to memcache + redis checks

bram123 opened the pull request:
Signed-off-by: Bram Leeda <bram@123inkt.nl>

|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Added more CheckResult data values:
- responsetime
- uptime
- connections

Added as a keyed array following DirWritable, or do you prefer a dataclass for this for structured data?
There were no unittests for these classes unfortunately, to test this I'd need to be able to mock the Client instance.

bram123 committed:
Add responsetime/uptime/connection-count to memcache + redis

Signed-off-by: Bram Leeda <bram@123inkt.nl>

bram123 commented:
If this change is acceptable, I'd also like to add these stats to the PDOCheck class.
Problem here is that in order to do this, I'd need to execute a query. And I know the syntax for MySQL, but I don't know a generic method to fetch this.
Is a mysql specific implementation for this allowed (and empty result-data for other drivers)?

Ocramius commented on src/Check/Memcache.php:
```diff
@@ -43,7 +44,7 @@ public function __construct($host = '127.0.0.1', $port = 11211)
         if ($port < 1) {
             throw new InvalidArgumentException(sprintf(
-                'Invalid port number - expecting a positive integer',
+                'Invalid port number %s - expecting a positive integer',
```
This code didn't make much sense upfront: `$host` is not part of this error, `$port` has always been cast to integer :thinking:

bram123 replied:
Ah yes ofc, altered the error and added UT for the constructor validation specifically. Had to update psalm-baseline though because I'm sending an incorrect $host type.

Ocramius commented on src/Check/Memcache.php:
```diff
@@ -66,13 +67,17 @@ public function check()
 
-            $authority = sprintf('%s:%d', $this->host, $this->port);
+            $startTime = microtime(true);
+            /** @var false|array<string, false|array<string, int|string>> $stats */
```
Would it make sense to introduce a `MemcacheService.php.stub` in the project? That would de-clutter the source here :thinking:

bram123 replied:
Opened a PR to psalm instead: https://github.com/vimeo/psalm/pull/8700 updating the types on their end.

Ocramius replied:
Excellent, thanks!

Ocramius commented on src/Check/Memcache.php:
```diff
@@ -84,6 +89,12 @@ public function check()
                     ));
                 }
+            } else {
+                $serviceData = [
+                    "responseTime" => $responseTime,
+                    "connections"  => (int) $stats[$authority]['curr_connections'],
+                    "uptime"       => (int) $stats[$authority]['uptime'],
+                ];
```
Can we perhaps make the type more precise, by having a `ServiceData<{responseTime: float, connections: int, uptime: int}>` or such? :thinking:

Ocramius commented on src/Check/Memcached.php:
```diff
@@ -43,7 +44,7 @@ public function __construct($host = '127.0.0.1', $port = 11211)
         if ($port < 1) {
             throw new InvalidArgumentException(sprintf(
-                'Invalid port number - expecting a positive integer',
+                'Invalid port number %s - expecting a positive integer',
```
Similar to above: we check the port, then interpolate the type of the **host**? The error message also reads very weirdly here

Ocramius commented on src/Check/Memcached.php:
```diff
@@ -66,9 +67,14 @@ public function check()
 
-            $authority = sprintf('%s:%d', $this->host, $this->port);
+            $startTime = microtime(true);
+            /** @var false|array<string, false|array<string, int|string>> $stats */
```
The same feedback of `Memcache.php` also applies to `Memcached.php`

bram123 replied:
Opened a PR to psalm instead: https://github.com/vimeo/psalm/pull/8700 updating the types on their end.

Ocramius reviewed (changes requested):
Proposed some suggestions: expanding on the success data is very viable, but let's try to improve on the API with this :+1:

bram123 committed:
Fix Memcache(d) port number validation error
Added unittest for Memcache(d) constructor's host/port validation

Had to update psalm-baseline because we are testing invalid input for $host

Signed-off-by: Bram Leeda <bram@123inkt.nl>

bram123 committed:
Update psalm-baseline, add extension baselines that were removed after locally installing the extensions

Signed-off-by: Bram Leeda <bram@123inkt.nl>

Ocramius reviewed (approved):
LGTM here.

:ship: meanwhile, good improvement!
